### ISSUE TITLE
Work out a way to declaratively create new cases without required a new data type

### ISSUE THREAD

**Issue by louispan** [louispan/data-distinct#6]
Eg. avoid CaseTypeable, CaseNum, etc
look at https://hackage.haskell.org/package/polydata-0.1.0.0/docs/Data-Poly.html ?

**Comment by louispan** [comment]
It requires UndecidableSuperClasses :(

**Comment by louispan** [comment]
Solution courtesy of Syrak
https://www.reddit.com/r/haskell/comments/75zrci/help_how_to_pass_constraints_as_a_type_variable/

**Comment by clintonmead** [comment]
I thought I'd reply to your email in this thread so it's easier to track.

I had earlier thought about splitting up [polydata](https://hackage.haskell.org/package/polydata) so now I have, there's now [polydata-core](https://hackage.haskell.org/package/polydata-core) which should be useful enough for your purposes.

It's core module doesn't use `UndecidableSuperClasses`, although to implement `CaseFunc` you I haven't worked out a way to avoid it. Here's the code:

```
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE UndecidableSuperClasses #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE UndecidableInstances #-}

module Main where

import Data.Poly (Poly)

import Data.Kind (Type)
import GHC.Exts (Constraint)

type family Arg a where
  Arg (a -> _) = a

type family Result a where
  Result (_ -> a) = a

class (a ~ (Arg a -> Result a), Arg a ~ Result a, k a) => ConstrainedFunc k a
instance (a ~ (Arg a -> Result a), Arg a ~ Result a, k a) => ConstrainedFunc k a

type CaseFunc (k :: Type -> Constraint) = Poly (ConstrainedFunc k)

main :: IO ()
main = pure ()
```

What's actually requiring `UndecidableSuperClasses` is the `k a` part, which I don't think can be avoided in your use case. I don't think `UndecidableSuperClasses` is that scary though, all enabling the extension does as far as I know is allow you to produce code that will result in the compiler looping forever (much like `UndecidableInstances`). I don't think it does anything particularly scary (unlike `IncoherentInstances` which can be scary).

**Comment by clintonmead** [comment]
Also, just a quick question, can your library do "fast" (less than linear than the current length of the structure) inserts and deletes in the middle of the structure, or only at the ends?

**Comment by louispan** [comment]
Hmm, at the moment, I don't provide an api to insert in the middle of the structure, but I could add one if you require it. SInce the underlying data structure is a `Seq [Any]`, I can use Seq's [insertAt](https://hackage.haskell.org/package/containers-0.5.10.1/docs/Data-Sequence.html) which is O(log(min(i,n-i)))

As for deleting, I also don't provide an explicit api to delete, but you can use [select](https://github.com/louispan/data-diverse/blob/master/test/Data/Diverse/ManySpec.hs#L226) to select all but the delete field.

**Comment by louispan** [comment]
@clintonmead I've created https://github.com/louispan/data-diverse/issues/7 to track this. I'll add it this weekend.

**Comment by louispan** [comment]
@clintonmead I've upload data-diverse-1.2.0.1 with split & splice in the middle.
There is also insert/remove individual items. However, insert/remove is only available in ghc 8.2.

**Comment by louispan** [comment]
I've added an implementation for insert/remove for all ghc versions in v1.2.0.0
http://hackage.haskell.org/package/data-diverse-1.2.0.2/docs/Data-Diverse-Many.html#g:17
